File: frontend/src/api/automation.ts
/**
 * Automation API client for managing scheduled jobs.
 */
import axios from './axios'

export interface Job {
  job_id: string
  name: string
  next_run_time: string | null
  trigger?: string
  enabled: boolean
}

export interface JobListResponse {
  jobs: Job[]
  total: number
}

export interface ScheduleBudgetAlertsRequest {
  job_id?: string
  cron_expression?: string
  enabled?: boolean
}

export interface ScheduleAuditRequest {
  job_id: string
  account_name: string
  cron_expression?: string
  audit_types?: string[]
  send_teams_notification?: boolean
  webhook_id?: number | null
  enabled?: boolean
}

export interface SchedulerStatus {
  running: boolean
  total_jobs: number
  active_jobs: number
}

export const automationApi = {
  /**
   * List all scheduled jobs.
   */
  async listJobs(): Promise<JobListResponse> {
    const response = await axios.get('/automation/jobs')
    return response.data
  },

  /**
   * Get details of a specific job.
   */
  async getJob(jobId: string): Promise<Job> {
    const response = await axios.get(`/automation/jobs/${jobId}`)
    return response.data
  },

  /**
   * Schedule automated budget alerts.
   */
  async scheduleBudgetAlerts(request: ScheduleBudgetAlertsRequest = {}): Promise<Job> {
    const response = await axios.post('/automation/budget-alerts/schedule', request)
    return response.data
  },

  /**
   * Schedule automated audit job.
   */
  async scheduleAudit(request: ScheduleAuditRequest): Promise<Job> {
    const response = await axios.post('/automation/audits/schedule', request)
    return response.data
  },

  /**
   * Pause a scheduled job.
   */
  async pauseJob(jobId: string): Promise<{ success: boolean; message: string }> {
    const response = await axios.post(`/automation/jobs/${jobId}/pause`)
    return response.data
  },

  /**
   * Resume a paused job.
   */
  async resumeJob(jobId: string): Promise<{ success: boolean; message: string }> {
    const response = await axios.post(`/automation/jobs/${jobId}/resume`)
    return response.data
  },

  /**
   * Delete a scheduled job.
   */
  async deleteJob(jobId: string): Promise<{ success: boolean; message: string }> {
    const response = await axios.delete(`/automation/jobs/${jobId}`)
    return response.data
  },

  /**
   * Run a job immediately (doesn't affect schedule).
   */
  async runJobNow(jobId: string): Promise<{ success: boolean; message: string }> {
    const response = await axios.post(`/automation/jobs/${jobId}/run-now`)
    return response.data
  },

  /**
   * Get scheduler status.
   */
  async getSchedulerStatus(): Promise<SchedulerStatus> {
    const response = await axios.get('/automation/status')
    return response.data
  },
}

File: backend/app/schemas/automation.py
"""
Pydantic schemas for Automation API endpoints.

These models define the request/response structure for:
- Scheduled job management
- Budget alert scheduling
- Audit scheduling
"""
from typing import List, Optional
from pydantic import BaseModel, Field


class ScheduleBudgetAlertsRequest(BaseModel):
    """Request to schedule budget alerts."""
    job_id: str = Field(
        default="budget-alerts-default",
        description="Unique job identifier"
    )
    cron_expression: str = Field(
        default="0 */6 * * *",
        description="Cron expression (default: every 6 hours)"
    )
    enabled: bool = Field(
        default=True,
        description="Enable job immediately after creation"
    )


class ScheduleAuditRequest(BaseModel):
    """Request to schedule audit job."""
    job_id: str = Field(
        ...,
        description="Unique job identifier"
    )
    account_name: str = Field(
        ...,
        description="AWS account to audit"
    )
    cron_expression: str = Field(
        default="0 2 * * *",
        description="Cron expression (default: daily at 2 AM)"
    )
    audit_types: Optional[List[str]] = Field(
        default=None,
        description="Audit types to run (null = all types)"
    )
    send_teams_notification: bool = Field(
        default=True,
        description="Send Teams notification when complete"
    )
    webhook_id: Optional[int] = Field(
        default=None,
        description="Specific webhook ID (null = all active webhooks)"
    )
    enabled: bool = Field(
        default=True,
        description="Enable job immediately after creation"
    )


class JobResponse(BaseModel):
    """Job details response."""
    job_id: str
    name: str
    next_run_time: Optional[str] = None
    trigger: Optional[str] = None
    enabled: bool


class JobListResponse(BaseModel):
    """List of jobs response."""
    jobs: List[JobResponse]
    total: int


class JobStatusResponse(BaseModel):
    """Job status response with progress details."""
    job_id: str
    status: str
    progress: int
    current_step: Optional[str] = None
    started_at: Optional[str] = None
    completed_at: Optional[str] = None
    error: Optional[str] = None
    result: Optional[dict] = None